File: Emulator/src/main/core/net/ServerSocketConnection.java
package core.net;

import java.nio.channels.Selector;
import java.nio.channels.ServerSocketChannel;
import java.nio.channels.SocketChannel;

/**
 * The type Server socket connection.
 */
public final class ServerSocketConnection {

    private final Selector selector;

    private final ServerSocketChannel channel;

    private final SocketChannel socket;

    /**
     * Instantiates a new Server socket connection.
     *
     * @param selector the selector
     * @param channel  the channel
     */
    public ServerSocketConnection(Selector selector, ServerSocketChannel channel) {
        this.selector = selector;
        this.channel = channel;
        this.socket = null;
    }

    /**
     * Instantiates a new Server socket connection.
     *
     * @param selector the selector
     * @param channel  the channel
     */
    public ServerSocketConnection(Selector selector, SocketChannel channel) {
        this.selector = selector;
        this.socket = channel;
        this.channel = null;
    }

    /**
     * Is client boolean.
     *
     * @return the boolean
     */
    public boolean isClient() {
        return socket != null;
    }

    /**
     * Gets selector.
     *
     * @return the selector
     */
    public Selector getSelector() {
        return selector;
    }

    /**
     * Gets channel.
     *
     * @return the channel
     */
    public ServerSocketChannel getChannel() {
        return channel;
    }

    /**
     * Gets socket.
     *
     * @return the socket
     */
    public SocketChannel getSocket() {
        return socket;
    }

}

File: Server/src/main/core/net/ServerSocketConnection.kt
package core.net

import java.nio.channels.Selector
import java.nio.channels.ServerSocketChannel
import java.nio.channels.SocketChannel

/**
 * Represents a server socket and its selector.
 *
 * @author Emperor
 */
class ServerSocketConnection {

    val selector: Selector
    val channel: ServerSocketChannel?
    val socket: SocketChannel?

    /**
     * Constructor for server socket channel.
     */
    constructor(selector: Selector, channel: ServerSocketChannel) {
        this.selector = selector
        this.channel = channel
        this.socket = null
    }

    /**
     * Constructor for client socket channel.
     */
    constructor(selector: Selector, socket: SocketChannel) {
        this.selector = selector
        this.socket = socket
        this.channel = null
    }

    /**
     * Returns true if this connection represents a client socket.
     */
    val isClient: Boolean
        get() = socket != null
}
